### ISSUE TITLE
How to cache a fragment based on last modified date

### ISSUE THREAD

**Issue by thejahweh** [yiisoft/yii2#8642]
I just want something like this:

``` php
$expiresOnThisValue = date('Y-m-d H:i:s');
if ($this->beginCache('test', ['expiresOnThisValues' => [
    $expiresOnThisValue
]])) {
    echo $expiresOnThisValue. "\n";
    $this->endCache();
}
```

So every time the `$expiresOnThisValue` has changed, the cache expires and the content gets newly loaded :tada:

I tried to write a dependency class with just a variable which is returned on `generateDependencyData`.

``` php
class VarDependency extends Dependency
{
    public $var;

    /**
     * @param \yii\caching\Cache $cache
     * @return mixed
     * @throws InvalidConfigException
     */
    protected function generateDependencyData($cache)
    {
        if ($this->var === null) {
            throw new InvalidConfigException('VarDependency::var must be set.');
        }
        return $this->var;
    }
}
```

My test was as simple like this:

``` php
$expiresOnThisValue = date('Y-m-d H:i:s');
if ($this->beginCache('test', ['dependency' => [
    'class' => 'VarDependency',
    'var' => $expiresOnThisValue 
]])) {
    echo $expiresOnThisValue . "\n";
    $this->endCache();
}
```

But then I realized, that the dependency class is cached too and does not set the var every time.

**Comment by mdmunir** [comment]
????
why not using cache duration = 1 second?

**Comment by kidol** [comment]
```php
if ($this->beginCache("test", ['variations' => [$date]])) {
```
http://www.yiiframework.com/doc-2.0/yii-widgets-fragmentcache.html#$variations-detail

**Comment by thejahweh** [comment]
@mdmunir because I want to cache it as long as `$expiresOnThisValue` changes, which can be everything, but in my case it is a last-modified date. This date changes if someone modified something and then this person wants to check his modification immediately.

@kidol the problem with `variations` is, that every variation is saved, so if you use the `$date` whit it, then there are hundreds of cache files, which will never expire.

**Comment by mdmunir** [comment]
```php
if ($this->beginCache('test', [
    'dependency' => [
        'class'=>'yii\caching\ExpressionDependency',
        'expression' => "date('Y-m-d H:i:s')",
    ]]){
    // your code
}
```

**Comment by thejahweh** [comment]
@mdmunir Sorry for that confusion, but its not an expression, its a variable which comes from elsewhere. My explanation with `date('Y-m-d H:i:s')` was just for testing.

**Comment by mdmunir** [comment]
whatever lah... just change `expresion` with what you want... 
You can also use `TagDependency`, `DbDependency`, etc, etc...

**Comment by thejahweh** [comment]
I think it is just not possible in the current state, because the dependency gets cached too. So the dependency doens't have a dynamic variable which could be used to check against the cached variable.

**Comment by kidol** [comment]
Try expression dependency with:
```php
'expression' => 'Yii::$app->controller->dateVar'
```

**Comment by thejahweh** [comment]
I've managed to do it with a change in the `FragmentCache` class:
```php
class FragmentCacheExtension extends FragmentCache 
{
	/**
	 * @var mixed
	 */
	public $dependentValue;

    public function run()
    {
        if (($content = $this->getCachedContent()) !== false) {
            echo $content;
        } elseif ($this->cache instanceof Cache) {
            $content = ob_get_clean();
            array_pop($this->getView()->cacheStack);
            if (is_array($this->dependency)) {
                $this->dependency = Yii::createObject($this->dependency);
            }
			// MAGIC START
            $data = [$content, $this->dynamicPlaceholders, $this->dependentValue];
			// MAGIC END
            $this->cache->set($this->calculateKey(), $data, $this->duration, $this->dependency);

            if (empty($this->getView()->cacheStack) && !empty($this->dynamicPlaceholders)) {
                $content = $this->updateDynamicContent($content, $this->dynamicPlaceholders);
            }
            echo $content;
        }
    }

    public function getCachedContent()
    {
        if ($this->_content === null) {
            $this->_content = false;
            if ($this->cache instanceof Cache) {
                $key = $this->calculateKey();
                $data = $this->cache->get($key);
				// MAGIC START
                if (is_array($data) && count($data) === 3) {
                    list ($content, $placeholders, $dependentValue) = $data;
					if ($dependentValue === $this->dependentValue) {
						// MAGIC END
						if (is_array($placeholders) && count($placeholders) > 0) {
							if (empty($this->getView()->cacheStack)) {
								// outermost cache: replace placeholder with dynamic content
								$content = $this->updateDynamicContent($content, $placeholders);
							}
							foreach ($placeholders as $name => $statements) {
								$this->getView()->addDynamicPlaceholder($name, $statements);
							}
						}
						$this->_content = $content;
					}
                }
            }
        }

        return $this->_content;
    }
}
```

I've added the `dependentValue` parameter, cache it in `run` and check it in `getCachedContent` if he changed since the last time. The `dependentValue` parameter, could be anything.

So the test would look like this:
```php
$date = date('Y-m-d H:i:s');
if ($this->beginCache('test', ['dependentValue' => [
    'a' // change manually to check if it works
]])) {
    echo $date. "\n";
    $this->endCache();
}
echo $date. "\n";
```

**Comment by thejahweh** [comment]
Should I make a commit with this change?
Or are there plans to implement it?

**Comment by OndrejVasicek** [comment]
This is absolutely great thejahweh. I was seeking for the same thing and was pretty confused that there is nothing like that. For a while I thought I can use ExpressionDependency but was very confused when it didn’t work, because even the whole dependency was cached. Then I successfully used variations, but realized, that even if it worked it’s definitely a bed idea to use it in this way, because it’s wasn’t meant. The side effect is that I’d multiple the cache for every different variant and the cache would bloat.

So your solution is absolutely simple and useful.
But wouldn’t be better to do it as a dependency? Because then we can use it anywhere – data, fragment, page caching. Now I had to extend my parent view and overwrite two methods – beginCache and endCache to force them using your extended class and it’s only usable in fragment caching.

If you were able to do it and make a PR, I’d be great.
